SU2_CFD/src/output/filewriter/CCGNSFileWriter.cpp

```diff
@@ -0,0 +1,310 @@
+ * License along with SU2. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "../../../include/output/filewriter/CCGNSFileWriter.hpp"
+
+const string CCGNSFileWriter::fileExt = ".cgns";
+
+CCGNSFileWriter::CCGNSFileWriter(string valFileName, CParallelDataSorter* valDataSorter, bool isSurf)
+    : CFileWriter(std::move(valFileName), valDataSorter, fileExt) {
+  isSurface = isSurf;
+}
+
+CCGNSFileWriter::~CCGNSFileWriter() {}
+
+void CCGNSFileWriter::Write_Data() {
+#ifdef HAVE_CGNS
+  /*--- Open the CGNS file for writing.  ---*/
+  initializeMeshFile();
+
+  /*--- Write point coordinates. ---*/
+  WriteCoordinate(1, "CoordinateX");
+  WriteCoordinate(2, "CoordinateY");
+  if (nDim == 3) WriteCoordinate(3, "CoordinateZ");
+
+  /*--- Write mesh connectivity. ---*/
+  if (nDim == 2) {
+    WriteConnectivity(LINE, "Lines");
+    WriteConnectivity(TRIANGLE, "Triangles");
+    WriteConnectivity(QUADRILATERAL, "Quadrilaterals");
+  }
+  if (nDim == 3) {
+    WriteConnectivity(TRIANGLE, "Triangles");
+    WriteConnectivity(QUADRILATERAL, "Quadrilaterals");
+    WriteConnectivity(TETRAHEDRON, "Tetrahedra");
+    WriteConnectivity(PYRAMID, "Pyramids");
+    WriteConnectivity(PRISM, "Prisms");
+    WriteConnectivity(HEXAHEDRON, "Hexahedra");
+  }
+
+  /*--- Initialize and write fields. ---*/
+  InitializeFields();
+
+  const vector<string> fieldNames = dataSorter->GetFieldNames();
+  for (unsigned long i = nDim; i < fieldNames.size(); ++i) {
+    WriteField(i, fieldNames[i]);
+  }
+
+  /*--- Close the CGNS file. ---*/
+  if (rank == MASTER_NODE) CallCGNS(cg_close(cgnsFileID));
+
+#endif
+}
+
+#ifdef HAVE_CGNS
+void CCGNSFileWriter::initializeMeshFile() {
+  if (!dataSorter->GetConnectivitySorted()) {
+    SU2_MPI::Error("Connectivity must be sorted.", CURRENT_FUNCTION);
+  }
+
+  int nCell;
+
+  nLocalPoints = static_cast<int>(dataSorter->GetnPoints());
+  nDim = dataSorter->GetnDim();
+  GlobalElem = static_cast<cgsize_t>(dataSorter->GetnElemGlobal());
+  GlobalPoint = static_cast<cgsize_t>(dataSorter->GetnPointsGlobal());
+  cumulative = 0;
+
+  /*--- If surface file cell dimension is decreased. ---*/
+  nCell = nDim - isSurface;
+
+  if (rank == MASTER_NODE) {
+    /*--- Remove the previous file if present. ---*/
+    remove(fileName.c_str());
+
+    /*--- Create CGNS file and open in write mode. ---*/
+    CallCGNS(cg_open(fileName.c_str(), CG_MODE_WRITE, &cgnsFileID));
+
+    /*--- Create Base. ---*/
+    CallCGNS(cg_base_write(cgnsFileID, "Base", nCell, nDim, &cgnsBase));
+
+    /*--- Create Zone. ---*/
+    vector<cgsize_t> zoneData(3);
+
+    zoneData[0] = GlobalPoint;
+    zoneData[1] = GlobalElem;
+    zoneData[2] = 0;
+
+    CallCGNS(cg_zone_write(cgnsFileID, cgnsBase, "Zone", zoneData.data(), Unstructured, &cgnsZone));
+  }
+}
+
+void CCGNSFileWriter::WriteCoordinate(int CoordinateNumber, const string& CoordinateName) {
+  cgsize_t nodeBegin;
+  cgsize_t nodeEnd;
+  vector<float> sendBufferFloat, recvBufferFloat;
```

Comment:
You could make the buffers class members and allocate them only once (using the max-size for the receive buffer) to avoid re-allocs.
Otherwise declare them where you use them.